### ISSUE TITLE
Deprecate var, invoke, assign

### ISSUE THREAD

**Issue by mlrawlings** [marko-js/marko#558]
With the new `$ <code>` syntax, we have a better way to add JS to a template and no longer need these tags.  We do want to make them available for the v3 -> v4 migration.  Should we just add a deprecation warning or split them out into a `marko-v3-compat` package?

**Comment by mlrawlings** [comment]
My preference would be to split them out so we're rid of them as v4 users, but it makes things a little tricky with sub packages that export tags because taglibs outside of Marko are not global.

**Comment by mindeavor** [comment]
I think a migration package is a good idea. If one intends to move from v3 to v4, installing a package would be the least of those efforts.

What do you mean by sub packages that export tags?

**Comment by mlrawlings** [comment]
@mindeavor Say I am using a package with a taglib that is intended to be share across apps, let's call it `ui-components` and say one of the tags it exports is `<ui-carousel>`.  

`ui-components` was developed against v3 and as such uses legacy widgets.  The way things currently stand with v4, these components should still be able to be used with a v4 app.

The problem in moving out to a package is that if the `<ui-carousel>` template uses the `<var>` tag, it wouldn't be discovered because templates only look for tags in packages that their package (in this case `ui-components`) directly depends on.  It wouldn't look at the application's dependencies to find `marko-v3-compat`.

**Comment by mlrawlings** [comment]
This really isn't an issue as we could use the not-really-documented [`taglibLookup.registerTaglib`](https://github.com/marko-js/marko/blob/master/compiler/index.js#L229-L234) to make it globally available.  But if there is any runtime code (which there isn't in this case) using that API can cause discoverability problems.

**Comment by Hesulan** [comment]
@mlrawlings I think this really comes down to the developer knowing that using a v3 taglib in v4 can be tricky. These tags aren't the only breaking changes; The switch from `-` to `--` in concise comes to mind, along with several other edge-cases. Unless the parser and compiler can be switched into some sort of "legacy" mode whenever an outdated taglib is being loaded, I don't see any other way around the occasional compatibility issue, though in some cases a `marko-v3-compat` taglib might help.

**Comment by mlrawlings** [comment]
@Hesulan We actually are using the parse in legacy mode in v3 and have a number of warnings, including the [switch from `-` to `--`](https://github.com/marko-js/htmljs-parser/blob/master/Parser.js#L1359-L1361).  The goal is to allow users to upgrade without any changes if there are no deprecation warnings in v3.

**Comment by mindeavor** [comment]
This might be a crazy idea, but what about a `marko-v4-stepping-stone` package that enables v4 syntax for v3 and adds deprecation warnings for all the old stuff from v3?

**Comment by ramses0** [comment]
As a recreational user of marko v3 transitioning to v4, I am 100% in favor of "kill stuff that's going to die" (ie: `var`, etc).  `npm install markojs-v3-deprecated` is totally fine by me.

I *will* say that I am especially disappointed by the removal of the `<var>...</var>` special tag.

Conceptually being able to scope my sins of using `<var/>` in my logicless templates in the first place is invaluable.  Forgive my use of tables (throwaway code during development), but I used them like this:

```
      <var
        flavors=data.fnRandomFlavors( data.selection )
        ingredients=data.fnRandomPopularIngredients()
      >
        <h1>${data.fnCap(data.selection)} and...</h1>
             <td align="right"><a href="/drinks-with/${data.selection}/$${flavors[0]}">
              ${data.fnCap(flavors[0])}
            </a></td>
            <td>or</td>
            <td><a href="/drinks-with/${data.selection}/$${flavors[1]}">
              ${data.fnCap(flavors[1])}
            </a></td>
          </tr>
          <tr>
            <td align="right"><a href="/drinks-with/${data.selection}/${ingredients[0]}">
              ${data.fnCap(ingredients[0])}
            </a></td>
            <td>or</td>
            <td><a href="/drinks-with/${data.selection}/${ingredients[1]}">
              ${data.fnCap(ingredients[1])}
            </a></td>
          </tr>
```


The introduction of `$ var flavor=input.someFunc()` kill the template and you end up with PHP all over again.

What I mean is that with a scoped `<var>` tag, I can easily determine if the template is factorable (ie: what dependencies or requirements it might have).

With a "dangling `$ var ...`" statement... Every. Single. Line. after that `$ var ...` statement is suspect, and requires brainpower to determine if it is using any of the declared `$ var ...` variables.

In summary:

1) marko-v3-compat package is totally fine

2) killing `<var />`, or encouraging the use of `$ ...` for anything other than debugging is bordering on criminal (it basically makes the entire file unable to be reasoned about unless the human first does a "compilation pass" looking for `$ var ...` and then remembering where it is used).

I've written and seen enough bad PHP, and bad JS, written by good people (but not good programmers) that I would view use of markojs as highly suspicious.

*THINK* about this.  When you see `Array.prototype.something = function() { ... }` does that make you happy that the person before you was so clever? Or fill you with rage that now you have to audit the entire source tree for uses of `*.prototype` and then see how that interacts with the rest of the codebase?

In the below code, `<var a=...>` is demonstrable better from a maintenance programmer perspective compared to `$ var b=...;`  Neither is ideal, but `<var a=...>` as a shortcut to make the layout of the template easier, cleaner, etc. is much better.

```
<div>
    <var a=fnFoo()>
        ...snip...
        Hello ${a}
    </var>
    $ var b=fnBar();
    ...snip...
    <div>
    ...snip...
    Hello ${b}
    </div>
</div>

**Comment by Hesulan** [comment]
@ramses0 I believe the main reason for deprecating `<var>`, `<assign>`, `<invoke>`, etc. is because the number of tags that attempt to simulate real javascript was becoming unmanageable, but you make a good point.

@mlrawlings Perhaps it would be a good idea to hold off on removing the `<var>` tag? I feel like it's enough of a special case to warrant keeping it around, especially since it fits each syntax so nicely.

**Comment by mindeavor** [comment]
Oh interesting, I didn't know `<var>` could use body text like that. Is that scope actually enforced?

**Comment by Hesulan** [comment]
@mindeavor Internally, a `<var>` tag with body text compiles to an IIFE, which enforces scope:
```html
<var foo=1 bar=2>
    /* ... */
</var>
```
```js
(function(){
    var foo=1, bar=2;
    /* ... */
})();
```

**Comment by mindeavor** [comment]
> The introduction of $ var flavor=input.someFunc() kill the template and you end up with PHP all over again.

Let's be honest here: We don't get PHP simply because of `<? ?>` tags. Unless you consider Ruby on Rails and ERB to be PHP as well!

I do find the `<var x=10>...</var>` feature interesting. To avoid potential confusion with JS, maybe you could rename it to something like `<scoped x=10>`, for those who want that feature.

**Comment by patrick-steele-idem** [comment]
Fixed by https://github.com/marko-js/marko/commit/768d8085f7adc8811e027ec7a88cae682a703a9e (thank you @austinkelleher !)

**Comment by ramses0** [comment]
Will there be (or is there) a way within marko to trigger an IIFE context for scope protection, as referenced in this comment?

https://github.com/marko-js/marko/issues/558#issuecomment-277313323

**Comment by patrick-steele-idem** [comment]
The scoped variables use case is interesting. I'm not opposed to introducing something to support that use case, but I am not convinced `<var>...</var>` was the best solution. @ramses0, what are your thoughts on the following? :

```html
<div>
    <scope>
        $ var name = 'Frank';
        $ function doSomething() {
        
        }

        <span>Hello ${name}!</span>
    </scope>
    <!-- `name` will be undefined here -->
</div>
```

**Comment by ramses0** [comment]
<scope/> is elegantly nasty in a way.  :-) ... it makes it very clear what you're doing, but still requires you to use `$ ...` but then you know your `$ ...`'s can't infect outside that `<scope>` block.  I guess I'm highly in favor of it... good design instinct.  "it would be nice" if there was a setting to emit a warning of something like the following:

```
Warning: file.marko:123  raw javascript `$ ...` found not at top of file or top of <scope> block.
```

ie: encourage `$` items to be in blocks at the beginning of the file instead of randomly sprinkled throughout.  If I can consistently open a template and know that `$ ...` is all bunched up at the top then I'll be a much happier camper.  But adding in a `$ ...debugging()...` randomly might be perfectly appropriate and encouraged (just not checked in / not for prod).

In short, though... I really appreciate you taking into consideration the issue of `<scope>` and would be highly in favor of it!
